### PR TITLE
Anbangx/add bubble merge statistics

### PR DESCRIPTION
**Pull request by anbangx**
@jakebiesinger @JavierJia @Nan-Zhang @Elmira88 

This PR provides statistics counters and log information for bubble merge.

For statistics counters, you can see the distribution histograms by specifying command line option -runAllStats.  Distribution histograms include totalPaths, removedPaths, editDistance, fractionDissimilar, totalPathsCoverage and removedPathsCoverage.

For log information, you can get all relative information of bubbles by specifying command line option -logBubbleInfo. 
Ex. A and B are two paths to form bubble, C and D are the major and minor of AB bubble respectively.

```
LOG:
Major: C   Minor: D
numOfPaths: 2
    No.1 bubble(internalKmer): A   PathLength: ...  Coverage: ..    ID: ...
        Dissimilar: .... with B bubble
    No.2 bubble(internalKmer): B   PathLength: ...  Coverage: ..    ID: ...
        Dissimilar: .... with A bubble
Remove # bubbles: No.2 bubble - A;
///////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```

I merged with our current branch. Jake just updated MaxLength for bubble. I run it on ecoli dataset in hpc. You can check the log under ~anbangx/workerlog and the distribution histograms under /cbcl/anbangx/output/bubble-stats/01-BUBBLE-STATS/counters.

### CODE REVIEW

File: genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/config/GenomixJobConf.java [hunk 1]
```diff
@@ -171,7 +171,9 @@
 
         @Option(name = "-runAllStats", usage = "Whether or not to run a STATS job after each normal job")
         private boolean runAllStats = false;
-
+        
+        @Option(name = "-logBubbleInfo", usage = "Whether or not to log bubble info on simple bubble merge")
```
Comment (jakebiesinger) [inline genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/config/GenomixJobConf.java:175]: Why don't we just use `-debugKmers`?  If you specify `-debugKmers ""`, all kmers will be "logged" in each step and you can decide what things you want to log.  That would alos let you specify a subset of the bubbles to inspect, just like we do for pathmerge.

File: genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/types/VKmer.java [hunk 2]
```diff
@@ -820,6 +820,17 @@ public float fracDissimilar(boolean sameOrientation, VKmer other) {
             return fracDissimilar(this, reverseKmer);
         }
     }
+    
+    public float editDistance(boolean sameOrientation, VKmer other) {
```
Comment (jakebiesinger) [inline genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/types/VKmer.java:824]: was this not exposed publicly before?

File: genomix/genomix-hadoop/src/main/java/edu/uci/ics/genomix/hadoop/utils/GraphStatistics.java [hunk 3]
```diff
@@ -134,7 +134,8 @@ public static Counters run(String inputPath, String outputPath, JobConf baseConf
     /**
      * run a map-reduce job on the given input graph and save a simple text file of the relevant counters
      */
-    public static void saveGraphStats(String outputDir, Counters jobCounters, GenomixJobConf conf) throws IOException {
+    public static void saveGraphStats(String outputDir, Counters jobCounters, GenomixJobConf conf, boolean runLocal)
```
Comment (jakebiesinger) [inline genomix/genomix-hadoop/src/main/java/edu/uci/ics/genomix/hadoop/utils/GraphStatistics.java]: you don't need to specify `local` or not-- instead, pass a `new Configuration()` object.

File: genomix/genomix-pregelix/data/TestSet/MERGE/SimpleBubble/graphviz.svg [hunk 4]
```diff
@@ -0,0 +1,213 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (jakebiesinger) [inline genomix/genomix-pregelix/data/TestSet/MERGE/SimpleBubble/graphviz.svg:1]: why'd you commit this?

File: genomix/genomix-pregelix/src/main/java/edu/uci/ics/genomix/pregelix/operator/simplebubblemerge/SimpleBubbleMergeVertex.java [hunk 5]
```diff
@@ -110,8 +114,68 @@ public static boolean isValidMajorAndMinor(SimpleBubbleMergeMessage topMsg, Simp
...
+            if (logBubbleInfo) {
+                logInfo += "\n";
+            }
+        }
+    }
+
     public void processSimilarSet() {
         while (!receivedMsgList.isEmpty()) {
+            /* log and update stats */
+            updateStatsAndLogBubbleInfo();
```
Comment (jakebiesinger) [inline genomix/genomix-pregelix/src/main/java/edu/uci/ics/genomix/pregelix/operator/simplebubblemerge/SimpleBubbleMergeVertex.java]: seems like this should have a big, fat `if (debug)` gating out this call.

File: genomix/genomix-pregelix/src/test/java/edu/uci/ics/genomix/pregelix/jobrun/pipeline/SimpleBubbleMergeTestSuite.java [hunk 6]
```diff
@@ -7,8 +7,7 @@
 
     public static Test suite() throws Exception {
         String pattern = "BUBBLE";
-        String testSet[] = { "TwoSimilarInTwo", "ThreeSimilarInThree", "TwoSimilarInThree", "TwoSimilarInFour",
-                "SimpleLongReads", "SimpleLongReads3" };
+        String testSet[] = { "SimpleBubble" };
```
Comment (jakebiesinger) [inline genomix/genomix-pregelix/src/test/java/edu/uci/ics/genomix/pregelix/jobrun/pipeline/SimpleBubbleMergeTestSuite.java:10]: why fewer test cases?

File: genomix/genomix-pregelix/src/test/java/edu/uci/ics/genomix/pregelix/jobrun/pipeline/PathMergeTestSuite.java [hunk 7]
```diff
@@ -7,12 +7,13 @@
     
     public static Test suite() throws Exception {
         String pattern = "MERGE";
-        String testSet[] = { "2", "3", "4", "5", "6", "7", "8", "9", "Head_5", "Head_6", "Head_7", "Head_8", "Head_9",
-                "Head_10", "LeftAdj", "RightAdj", "SimplePath", "ThreeDuplicate", "SimpleBridgePath", "SimpleTreePath",
-                "RingPath", "CyclePath", "SelfTandemRepeat", "TandemRepeatWithMergeEdge",
-                "TandemRepeatWithUnmergeEdge", "ComplexTandemRepeat", "MultiTandemRepeat", "MultiTandemRepeat2",
-                "MultiTandemRepeat3", "TandemRepeatWithThreeNodes", "TandemRepeatAndCycle", "ThreeNodesCycle",
-                "SameGeneCodeWithEdge", "SameGeneCodeWithoutEdge", "synthetic" };
+        String testSet[] = {"SimpleBubble"};
```
Comment (jakebiesinger) [inline genomix/genomix-pregelix/src/test/java/edu/uci/ics/genomix/pregelix/jobrun/pipeline/PathMergeTestSuite.java:10]: why fewer test cases?

**Comment by jakebiesinger** [comment]
I think the log line showing the node that's removed should have _both_ the `vertexId` and the `internalKmer`.  Otherwise, there will be some struggle grepping!

File: genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/config/GenomixJobConf.java (hunk 1 above)
Comment (anbangx) [inline genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/config/GenomixJobConf.java:175]: But we have another log for if(versebose), which records other stuff. Actually they are useless

File: genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/types/VKmer.java (hunk 2 above)
Comment (anbangx) [inline genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/types/VKmer.java:824]: we have editDistance(VKmer other) before, but not editDistance(boolean sameOrientation, VKmer other)

File: genomix/genomix-hadoop/src/main/java/edu/uci/ics/genomix/hadoop/utils/GraphStatistics.java (hunk 3 above)
Comment (anbangx) [inline genomix/genomix-hadoop/src/main/java/edu/uci/ics/genomix/hadoop/utils/GraphStatistics.java]: OK. I see.

File: genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/config/GenomixJobConf.java (hunk 1 above)
Comment (jakebiesinger) [inline genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/config/GenomixJobConf.java:175]: > Actually they are useless

What are useless?  Sorry, I don't follow.  You mean they aren't used in this algorithm?

That's exactly what I'm saying.  `-debugKmers` is our own debugging mechanism and can mean whatever we want it to mean.  In BubbleMerge, why don't why take it to mean what `-logBubbleInfo` means now?

Comment (anbangx) [inline genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/config/GenomixJobConf.java:175]: I mean our old log using if(verbose) is useless. Sure. We can remove  -logBubbleInfo and use  -debugKmers to log these useful logs

Comment (jakebiesinger) [inline genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/config/GenomixJobConf.java:175]: what do you mean it's useless?  You mean it's broken?

Comment (anbangx) [inline genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/config/GenomixJobConf.java:175]: It only records what major vertex is

File: genomix/genomix-hadoop/src/main/java/edu/uci/ics/genomix/hadoop/utils/GraphStatistics.java (hunk 3 above)
Comment (anbangx) [inline genomix/genomix-hadoop/src/main/java/edu/uci/ics/genomix/hadoop/utils/GraphStatistics.java]: Done

File: genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/config/GenomixJobConf.java (hunk 1 above)
Comment (jakebiesinger) [inline genomix/genomix-data/src/main/java/edu/uci/ics/genomix/data/config/GenomixJobConf.java:175]: Each algorithm is responsible for implementing logging.  If it doesn't do what you want it to, you should change it.  Maybe I'm misunderstanding what you mean :P

File: genomix/genomix-pregelix/data/TestSet/MERGE/SimpleBubble/graphviz.svg (hunk 4 above)
Comment (anbangx) [inline genomix/genomix-pregelix/data/TestSet/MERGE/SimpleBubble/graphviz.svg:1]: Always forget ... Remove it! Done

File: genomix/genomix-pregelix/src/test/java/edu/uci/ics/genomix/pregelix/jobrun/pipeline/PathMergeTestSuite.java (hunk 7 above)
Comment (anbangx) [inline genomix/genomix-pregelix/src/test/java/edu/uci/ics/genomix/pregelix/jobrun/pipeline/PathMergeTestSuite.java:10]: Fix!

**Comment by anbangx** [comment]
@jakebiesinger Change logBubbleInfo to verbose!

**Comment by anbangx** [comment]
And fixed GraphMutations.Num_RemovedBubbles.

**Comment by anbangx** [comment]
Can I merge this?

File: genomix/genomix-pregelix/src/main/java/edu/uci/ics/genomix/pregelix/operator/simplebubblemerge/SimpleBubbleMergeVertex.java (hunk 5 above)
Comment (jakebiesinger) [inline genomix/genomix-pregelix/src/main/java/edu/uci/ics/genomix/pregelix/operator/simplebubblemerge/SimpleBubbleMergeVertex.java]: Can you add the if statement here?

**Comment by jakebiesinger** [comment]
What does the log line look like after this change?

File: genomix/genomix-pregelix/src/main/java/edu/uci/ics/genomix/pregelix/operator/simplebubblemerge/SimpleBubbleMergeVertex.java (hunk 5 above)
Comment (anbangx) [inline genomix/genomix-pregelix/src/main/java/edu/uci/ics/genomix/pregelix/operator/simplebubblemerge/SimpleBubbleMergeVertex.java]: Done!

**Comment by anbangx** [comment]
@jakebiesinger  The log line looks the same as you see in my byobu.

**Comment by jakebiesinger** [comment]
LGTM.

**Pull request merged**